src/compiler/checker.ts

```diff
@@ -7157,6 +7183,17 @@ namespace ts {
                 propertiesArray.push(member);
             }
 
+            // If object literal is contextually typed by the implied type of a binding pattern, augment the result
+            // type with those properties for which the binding pattern specifies a default value.
+            if (contextualType && contextualType.flags & TypeFlags.ImpliedType) {

```

Comment:
It might be good to add tests that demonstrate that this can cause an error. Something like this, which I don't believe would have errored before:
```ts
function foo({ a = 0 } = {}) { }
foo({ a: "" });
```
Basically something where the property was dropped in the initializer, but then was present again in the argument with the wrong type.